packages/api/src/agent.ts

```diff
@@ -0,0 +1,298 @@
+
+  private _baseClient: AtpBaseClient
+  private _persistSession?: AtpPersistSessionHandler
+  private _refreshSessionPromise: Promise<void> | undefined
+
+  /**
+   * The `fetch` implementation; must be implemented for your platform.
+   */
+  static fetch: AptAgentFetchHandler | undefined = defaultFetchHandler
+
+  /**
+   * Configures the API globally.
+   */
+  static configure(opts: AtpAgentGlobalOpts) {
+    AtpAgent.fetch = opts.fetch
+  }
+
+  constructor(opts: AtpAgentOpts) {
+    this.service =
+      opts.service instanceof URL ? opts.service : new URL(opts.service)
+    this._persistSession = opts.persistSession
+
+    // create an ATP client instance for this agent
+    this._baseClient = new AtpBaseClient()
+    this._baseClient.xrpc.fetch = this._fetch.bind(this) // patch its fetch implementation
+    this.api = this._baseClient.service(opts.service)
+  }
+
+  /**
+   * Is there any active session?
+   */
+  get hasSession() {
+    return !!this.session
+  }
+
+  /**
+   * Sets the "Persist Session" method which can be used to store access tokens
+   * as they change.
+   */
+  setPersistSessionHandler(handler?: AtpPersistSessionHandler) {
+    this._persistSession = handler
+  }
+
+  /**
+   * Create a new account and hydrate its session in this agent.
+   */
+  async createAccount(
+    opts: AtpAgentCreateAccountOpts,
+  ): Promise<ComAtprotoAccountCreate.Response> {
+    try {
+      this.session = undefined
+      const res = await this.api.com.atproto.account.create({
+        handle: opts.handle,
+        password: opts.password,
+        email: opts.email,
+        inviteCode: opts.inviteCode,
+      })
+      this.session = {
+        accessJwt: res.data.accessJwt,
+        refreshJwt: res.data.refreshJwt,
+        handle: res.data.handle,
+        did: res.data.did,
+      }
+      return res
+    } finally {
+      if (this.session) {
+        this._persistSession?.('create', this.session)
+      } else {
+        this._persistSession?.('create-failed', undefined)
+      }
+    }
+  }
+
+  /**
+   * Start a new session with this agent.
+   */
+  async login(
+    opts: AtpAgentLoginOpts,
+  ): Promise<ComAtprotoSessionCreate.Response> {
+    try {
+      this.session = undefined
+      const res = await this.api.com.atproto.session.create({
+        identifier: opts.identifier,
+        password: opts.password,
+      })
+      this.session = {
+        accessJwt: res.data.accessJwt,
+        refreshJwt: res.data.refreshJwt,
+        handle: res.data.handle,
+        did: res.data.did,
+      }
+      return res
+    } finally {
+      if (this.session) {
+        this._persistSession?.('create', this.session)
+      } else {
+        this._persistSession?.('create-failed', undefined)
+      }
+    }
+  }
+
+  /**
+   * Resume a pre-existing session with this agent.
+   */
+  async resumeSession(
+    session: AtpSessionData,
+  ): Promise<ComAtprotoSessionGet.Response> {
+    this.session = session
```

Comment:
This seems a lot more natural than how the session client would have you do it 👍